### PR TITLE
feat(rpc): Replace TypedTransactionType

### PR DESCRIPTION
**Pull request by jenpaff**
Towards closing https://github.com/paradigmxyz/reth/issues/7490

Quick vibe check if I'm going the right way here

### CODE REVIEW

File: crates/rpc/rpc/src/eth/helpers/signer.rs [hunk 1]
```diff
@@ -103,16 +116,23 @@ impl EthSigner for DevSigner {
...
 
     use super::*;
 
     fn build_signer() -> DevSigner {
-        let addresses = vec![];
         let secret =
             SecretKey::from_str("4646464646464646464646464646464646464646464646464646464646464646")
                 .unwrap();
-        let accounts = HashMap::from([(Address::default(), secret)]);
+        // TODO: is there a better way to get the address from the secret key?
```
Comment (jenpaff) [inline crates/rpc/rpc/src/eth/helpers/signer.rs:129]: not sure if there's a better way

File: crates/rpc/rpc/src/eth/helpers/signer.rs [hunk 2]
```diff
@@ -221,4 +243,29 @@ mod tests {
...
+                data: Some(Bytes::from(message.to_vec())),
+                input: Some(Bytes::from(message.to_vec())),
+            },
+            nonce: Some(0u64),
+            ..Default::default()
+        };
+        let txn_signed = signer.sign_transaction(request, &from).await;
+        assert!(txn_signed.is_ok());
+
+        assert_eq!(Bytes::from(message.to_vec()), txn_signed.unwrap().input().0);
```
Comment (jenpaff) [inline crates/rpc/rpc/src/eth/helpers/signer.rs:269]: assert on signature

File: crates/rpc/rpc-eth-api/Cargo.toml [hunk 3]
```diff
@@ -21,7 +21,7 @@ reth-evm.workspace = true
 reth-primitives.workspace = true
 reth-provider.workspace = true
 reth-revm.workspace = true
-reth-rpc-types.workspace = true
+# reth-rpc-types.workspace = true
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/Cargo.toml:24]: cool, can we remove this entirely?

File: crates/rpc/rpc-eth-api/src/helpers/transaction.rs [hunk 4]
```diff
@@ -555,17 +391,20 @@ pub trait EthTransactions: LoadTransaction {
...
-                return match signer.sign_transaction(request, from) {
-                    Ok(tx) => Ok(tx),
-                    Err(e) => Err(e.into_eth_err()),
+        txn: TransactionRequest,
+    ) -> impl Future<Output = Result<TransactionSigned, Self::Error>> + Send {
+        async move {
+            let signers: Vec<_> = self.signers().read().iter().cloned().collect();
+            for signer in signers {
+                if signer.is_signer_for(from) {
+                    return match signer.sign_transaction(txn.clone(), from).await {
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/transaction.rs:400]: I don't think we need the clone here
```suggestion
                    return match signer.sign_transaction(txn, from).await {
```

File: crates/rpc/rpc-types/src/eth/transaction/typed.rs
Comment (mattsse) [inline crates/rpc/rpc-types/src/eth/transaction/typed.rs:1]: I assume we can get rid of this file entirely?

File: crates/rpc/rpc/src/eth/helpers/signer.rs [hunk 5]
```diff
@@ -79,18 +82,28 @@ impl EthSigner for DevSigner {
...
+        let signer = EthereumWallet::from(wallet);
 
-        Ok(TransactionSigned::from_transaction_and_signature(transaction, signature))
+        // build and sign transaction with signer
+        let txn_envelope: TxEnvelope =
+            request.build(&signer).await.map_err(|_| SignError::InvalidTransactionRequest)?;
+
+        // decode transaction into signed transaction type
+        let mut enveloped_tx = Vec::new();
+        txn_envelope.encode(&mut enveloped_tx);
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/helpers/signer.rs:102]: this can be `txn.encoded_2718()`

File: crates/rpc/rpc/src/eth/helpers/signer.rs [hunk 6]
```diff
@@ -79,18 +82,28 @@ impl EthSigner for DevSigner {
...
-        // convert to primitive transaction
-        let transaction =
-            to_primitive_transaction(request).ok_or(SignError::InvalidTransactionRequest)?;
-        let tx_signature_hash = transaction.signature_hash();
-        let signature = self.sign_hash(tx_signature_hash, *address)?;
+        // create local signer wallet from signing key
+        let sk_bytes = self.accounts.get(address).ok_or(SignError::NoAccount)?.secret_bytes();
+        let wallet =
+            PrivateKeySigner::from_bytes(&sk_bytes.into()).map_err(|_| SignError::NoAccount)?;
+        let signer = EthereumWallet::from(wallet);
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/helpers/signer.rs:94]: perhaps we can simplify this by changing SecretKey to

`PrivateKeySigner`?

https://github.com/paradigmxyz/reth/blob/5a2d892579dda42d3b2ab82836a7b8a38d8ec321/crates/rpc/rpc/src/eth/helpers/signer.rs#L32-L32

https://github.com/alloy-rs/alloy/blob/8b57b52f8c796ce785e4530bf9e6fc73ea743ecc/crates/signer-local/src/lib.rs#L36-L37

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, this looks pretty good already

some suggestions

File: crates/rpc/rpc-eth-api/Cargo.toml (hunk 3 above)
Comment (jenpaff) [reply crates/rpc/rpc-eth-api/Cargo.toml:24]: yes after I rebased seems like it's not used anymore at all- will have a go at removing entirely

File: crates/rpc/rpc/src/eth/helpers/signer.rs (hunk 5 above)
Comment (jenpaff) [reply crates/rpc/rpc/src/eth/helpers/signer.rs]: thanks ! updated

File: crates/rpc/rpc/src/eth/helpers/signer.rs [hunk 7]
```diff
@@ -41,23 +41,24 @@ impl DevSigner {
...
+    fn get_key(&self, account: Address) -> Result<&PrivateKeySigner> {
         self.accounts.get(&account).ok_or(SignError::NoAccount)
     }
 
     fn sign_hash(&self, hash: B256, account: Address) -> Result<Signature> {
-        let secret = self.get_key(account)?;
-        let signature = sign_message(B256::from_slice(secret.as_ref()), hash);
+        let sk = self.get_key(account)?.to_field_bytes();
+        let signature = sign_message(B256::from_slice(sk.as_ref()), hash);
         signature.map_err(|_| SignError::CouldNotSign)
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/helpers/signer.rs:62]: you can use the signer API here:

use alloy_signer::signer::SignerSync;

let sk = self.get_key(account)?.

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, one more suggestion

File: crates/rpc/rpc/src/eth/helpers/signer.rs (hunk 6 above)
Comment (jenpaff) [reply crates/rpc/rpc/src/eth/helpers/signer.rs]: updated now using signer api

File: crates/rpc/rpc-eth-api/Cargo.toml (hunk 3 above)
Comment (jenpaff) [reply crates/rpc/rpc-eth-api/Cargo.toml]: @mattsse I feel like this reth-rpc-types doesn't do much anymore, should I give it a go to remove it fully ? (in this or separate pr)?

File: crates/rpc/rpc/src/eth/helpers/signer.rs [hunk 8]
```diff
@@ -79,18 +80,25 @@ impl EthSigner for DevSigner {
...
-        request: TypedTransactionRequest,
+        request: TransactionRequest,
         address: &Address,
     ) -> Result<TransactionSigned> {
-        // convert to primitive transaction
-        let transaction =
-            to_primitive_transaction(request).ok_or(SignError::InvalidTransactionRequest)?;
-        let tx_signature_hash = transaction.signature_hash();
-        let signature = self.sign_hash(tx_signature_hash, *address)?;
+        // create local signer wallet from signing key
```
Comment (jenpaff) [inline crates/rpc/rpc/src/eth/helpers/signer.rs:88]: update comment

File: crates/rpc/rpc-types/src/lib.rs [hunk 9]
```diff
@@ -15,11 +15,13 @@ mod eth;
...
-pub use eth::transaction::{self, TypedTransactionRequest};
 #[cfg(feature = "jsonrpsee-types")]
 pub use eth::{
     engine,
     engine::{
         ExecutionPayload, ExecutionPayloadV1, ExecutionPayloadV2, ExecutionPayloadV3, PayloadError,
     },
 };
+
+use alloy_primitives as _;
```
Comment (jenpaff) [inline crates/rpc/rpc-types/src/lib.rs:26]: possibly could be removed can look into it

**Review by mattsse — APPROVED** [review]
nice, I only removed the unused deps and made sure the features are updated accordingly

File: crates/rpc/rpc-eth-api/Cargo.toml (hunk 3 above)
Comment (mattsse) [reply crates/rpc/rpc-eth-api/Cargo.toml]: yep: https://github.com/paradigmxyz/reth/issues/11198

**Pull request merged**